src/neptune/new/handler.py

```diff
@@ -287,6 +304,102 @@ def log(
+                For details, see https://docs.neptune.ai/api/universal/#wait
+
+        Examples:
+            >>> import neptune.new as neptune
+            >>> run = neptune.init_run()
+            >>> for epoch in range(n_epochs):
+            ...     ... # Your training loop
+            ...     run["train/epoch/loss"].append(loss)  # FloatSeries
+            ...     token = str(...)
+            ...     run["train/tokens"].append(token)  # StringSeries
+            ...     run["train/distribution"].append(plt_histogram, step=epoch)  # FileSeries
+        """
+        verify_type("step", step, (int, float, type(None)))
+        verify_type("timestamp", timestamp, (int, float, type(None)))
+        if step is not None:
+            step = [step]
+        if timestamp is not None:
+            timestamp = [timestamp]
+
+        value = ExtendUtils.validate_and_transform_to_extend_format(value)
+        self.extend(value, step, timestamp, wait, **kwargs)
+
+    def extend(
+        self,
+        values: ExtendDictT,
+        steps: Optional[Collection[float]] = None,
+        timestamps: Optional[Collection[float]] = None,
+        wait: bool = False,
+        **kwargs,
+    ) -> None:
+        """Logs a series of values by appending the provided collection of values to the end of the series.
+
+        Available for following series field types:
+
+            * `FloatSeries` - series of float values
+            * `StringSeries` - series of strings
+            * `FileSeries` - series of files
+
+        When you log the first value, the type of the value determines what type of field is created.
+        For more, see the field types documentation: https://docs.neptune.ai/api/field_types
+
+        Args:
+            values: Values to be added to the series field, as a dictionary or collection.
+            steps: Optional collection of indeces for the entries being appended. Must be strictly increasing.
+            timestamps: Optional collection of time indeces for the entries being appended, in Unix time format.
+                If None, the current time (obtained with `time.time()`) is used.
+            wait: If True, the client sends all tracked metadata to the server before executing the call.
+                For details, see https://docs.neptune.ai/api/universal/#wait
+
+        Example:
+            The following example reads a CSV file into a pandas DataFrame and extracts the values
+            to create a Neptune series.
+            >>> import neptune.new as neptune
+            >>> run = neptune.init_run()
+            >>> for epoch in range(n_epochs):
+            ...     df = pandas.read_csv("time_series.csv")
+            ...     ys = df["value"]
+            ...     ts = df["timestamp"]
+            ...     run["data/example_series"].extend(ys, timestamps=ts)
+        """
+        ExtendUtils.validate_values_for_extend(values, steps, timestamps)
```

Comment:
Calling `validate_values_for_extend` on every level makes `extend` and `append` to work in `O(n^2)` where `n` is depth of nested structure.
We can easily fix it, but I don't think it's a problem.